src/Provider/CreditMemoFileProvider.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\RefundPlugin\Provider;
+
+use Gaufrette\Exception\FileNotFound;
+use Sylius\RefundPlugin\Entity\CreditMemoInterface;
+use Sylius\RefundPlugin\Generator\CreditMemoFileNameGeneratorInterface;
+use Sylius\RefundPlugin\Generator\CreditMemoPdfFileGeneratorInterface;
+use Sylius\RefundPlugin\Manager\CreditMemoFileManagerInterface;
+use Sylius\RefundPlugin\Model\CreditMemoPdf;
+use Sylius\RefundPlugin\Repository\CreditMemoRepositoryInterface;
+use Webmozart\Assert\Assert;
+
+final class CreditMemoFileProvider implements CreditMemoFileProviderInterface
+{
+    public function __construct(
+        private CreditMemoRepositoryInterface $creditMemoRepository,
+        private CreditMemoFileNameGeneratorInterface $creditMemoFileNameGenerator,
+        private CreditMemoPdfFileGeneratorInterface $creditMemoPdfFileGenerator,
+        private CreditMemoFileManagerInterface $creditMemoFileManager,
+        private string $creditMemosPath,
+    ) {
+    }
+
+    public function provide(CreditMemoInterface $creditMemo): CreditMemoPdf
+    {
+        return $this->provideByCreditMemo($creditMemo);
+    }
+
+    public function provideById(string $creditMemoId): CreditMemoPdf
+    {
+        /** @var CreditMemoInterface|null $creditMemo */
+        $creditMemo = $this->creditMemoRepository->find($creditMemoId);
+        Assert::notNull($creditMemo);
+
+        return $this->provideByCreditMemo($creditMemo);
+    }
+
+    private function provideByCreditMemo(CreditMemoInterface $creditMemo): CreditMemoPdf
+    {
+        $fileName = $this->creditMemoFileNameGenerator->generateForPdf($creditMemo);
+
+        try {
+            $pdf = $this->creditMemoFileManager->get($fileName);
+        } catch (FileNotFound) {
+            $pdf = $this->creditMemoPdfFileGenerator->generate($creditMemo->getId());
+            $this->creditMemoFileManager->save($pdf);
+        }
```

Comment:
1. I'm not sure what moment are you asking about, creating credit memo or creating PDF?
2. It's interesting, because this kind of provider with a similar logic is your idea from InvoicingPlugin 😄 For me, there is a service for generating PDF file and this one is the one which is `on the higher level to react properly` if there is no PDF generated 😃 

As this feature is a new one provided here, there may be many credit memos that are created, but the PDF is not saved on the server, and because of that, this provider exists.

Reply:
Nobody is perfect, apparently, even me xD I'm 👍 for checking the existence of the credit memo and generating it if it's not there (I was thinking about the scenario you mentioned about past credit memo 🚀). My comment is based directly on the SRP - if there is "Provider" it should **provide** not **provide and optionally save**. That's why I still think this functionality (catching an exception and saving if necessary) should be introduced in the classes that are using this provider. 

There is another thing there, [setting the full path](https://github.com/Sylius/RefundPlugin/pull/373/files#diff-f5b61b848602facab683ba301b90ff050f6ba64afa0d8c1c21106f96b0e71c2bR61) which also kinda violates CQS+SRP pattern :dancer: As far as I see, this full path is only used in [the one place](https://github.com/Sylius/RefundPlugin/pull/373/files#diff-f3995f1b446c71df2aef5749ff9cf1531b2970932e4985854360b707afb84f2eR64), so maybe it's not needed at all? Some `FullPathResolver` would (hehe) solve the issue, or maybe even solve it with the private method in the sender itself (it seems as only sender needs this full path) 💥